File: src/main/java/club/deneb/client/features/modules/combat/Aura.java
package club.deneb.client.features.modules.combat;

import club.deneb.client.features.modules.misc.AutoTool;
import club.deneb.client.client.FriendManager;
import club.deneb.client.features.Category;
import club.deneb.client.features.Module;
import club.deneb.client.features.ModuleManager;
import club.deneb.client.utils.EntityUtil;
import club.deneb.client.utils.LagCompensator;
import club.deneb.client.value.Value;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Items;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.Vec3d;

/**
 * Credit : KAMI ,thank you.
 * Created by 086 on 12/12/2017.
 * Last Updated 5 August 2019 by hub
 */
@Module.Info(name = "Aura", category = Category.COMBAT, description = "Hits entities around you")
public class Aura extends Module {

    Value<Boolean> players = setting("Players", true);
    Value<Boolean> animals = setting("Animals", false);
    Value<Boolean> mobs = setting("Mobs", false);
    Value<Double> range = setting("Range", 5.5d,0.0d,10.0d);
    Value<Boolean> wait = setting("Wait", true);
    Value<Boolean> walls = setting("Walls", false);
    Value<Boolean> sharpness = setting("32k Switch", false);

    @Override
    public void onTick() {
        if (mc.player.isDead) {
            return;
        }
        boolean shield = mc.player.getHeldItemOffhand().getItem().equals(Items.SHIELD) && mc.player.getActiveHand() == EnumHand.OFF_HAND;
        if (mc.player.isHandActive() && !shield) {
            return;
        }

        if (wait.getValue()) {
            if (mc.player.getCooledAttackStrength(getLagComp()) < 1) {
                return;
            } else if (mc.player.ticksExisted % 2 != 0) {
                return;
            }
        }

        for (Entity target : Minecraft.getMinecraft().world.loadedEntityList) {
            if (!EntityUtil.isLiving(target)) {
                continue;
            }
            if (target == mc.player) {
                continue;
            }
            if (mc.player.getDistance(target) > range.getValue()) {
                continue;
            }
            if (((EntityLivingBase) target).getHealth() <= 0) {
                continue;
            }
            if (((EntityLivingBase) target).hurtTime != 0 && wait.getValue()) {
                continue;
            }
            if (!walls.getValue() && (!mc.player.canEntityBeSeen(target) && !canEntityFeetBeSeen(target))) {
                continue; // If walls is on & you can't see the feet or head of the target, skip. 2 raytraces needed
            }
            if (players.getValue() && target instanceof EntityPlayer && !FriendManager.isFriend(target.getName())) {
                attack(target);
                return;
            } else {
                if (EntityUtil.isPassive(target) ? animals.getValue() : (EntityUtil.isMobAggressive(target) && mobs.getValue())) {
                    if (ModuleManager.getModuleByName("AutoTool").isEnabled()) {
                        AutoTool.equipBestWeapon();
                    }
                    attack(target);
                    return;
                }
            }
        }
    }

    private boolean checkSharpness(ItemStack stack) {

        if (stack.getTagCompound() == null) {
            return false;
        }

        NBTTagList enchants = (NBTTagList) stack.getTagCompound().getTag("ench");

		if (enchants == null) {
			return false;
		}


        for (int i = 0; i < enchants.tagCount(); i++) {
            NBTTagCompound enchant = ((NBTTagList) enchants).getCompoundTagAt(i);
            if (enchant.getInteger("id") == 16) {
                int lvl = enchant.getInteger("lvl");
                if (lvl >= 16) {
                    return true;
                }
                break;
            }
        }

        return false;

    }

    private void attack(Entity e) {

        if (sharpness.getValue()) {

            if (!checkSharpness(mc.player.getHeldItemMainhand())) {

                int newSlot = -1;

                for (int i = 0; i < 9; i++) {
                    ItemStack stack = mc.player.inventory.getStackInSlot(i);
                    if (stack == ItemStack.EMPTY) {
                        continue;
                    }
                    if (checkSharpness(stack)) {
                        newSlot = i;
                        break;
                    }
                }

                if (newSlot != -1) {
                    mc.player.inventory.currentItem = newSlot;
                }

            }

        }

        mc.playerController.attackEntity(mc.player, e);
        mc.player.swingArm(EnumHand.MAIN_HAND);

    }

    private float getLagComp() {
        if (wait.getValue()) {
            return -(20 - LagCompensator.INSTANCE.getTickRate());
        }
        return 0.0F;
    }

    private boolean canEntityFeetBeSeen(Entity entityIn) {
        return mc.world.rayTraceBlocks(new Vec3d(mc.player.posX, mc.player.posX + mc.player.getEyeHeight(), mc.player.posZ), new Vec3d(entityIn.posX, entityIn.posY, entityIn.posZ), false, true, false) == null;
    }
}

File: src/main/java/club/deneb/client/features/modules/combat/Aura.kt
package club.deneb.client.features.modules.combat

import club.deneb.client.client.FriendManager
import club.deneb.client.features.Category
import club.deneb.client.features.Module
import club.deneb.client.features.modules.misc.AutoTool.Companion.equipBestWeapon
import club.deneb.client.utils.EntityUtil
import club.deneb.client.utils.Timer
import club.deneb.client.utils.WaitCounter
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Items
import net.minecraft.util.EnumHand
import net.minecraft.util.math.Vec3d
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.gameevent.TickEvent

/**
 * Updated by B_312 on 04/24/2021
 */
@Module.Info(
        name = "Aura",
        category = Category.COMBAT,
        description = "Hits entities around you"
)
class Aura : Module() {

    private val delayMode = setting("Mode", "Tick", listOf("Tick","Ms"))

    private val tickDelay = setting("TickDelay",2,0,100).m(delayMode,"Tick")
    private val msDelay = setting("MsDelay",100,0,5000).m(delayMode,"Ms")
    private val whileEating = setting("While Eating", true)
    private val multiTarget = setting("Multi Targets", false)
    private val attackPlayers = setting("Players", true)
    private val attackMobs = setting("Mobs", false)
    private val passive = setting("Passive Mobs",false).v { attackMobs.value }
    private val neutral = setting("Neutral Mobs",false).v { attackMobs.value }
    private val hostile = setting("Hostile Mobs",false).v { attackMobs.value }
    private val hitRange = setting("Hit Range", 5.5,0.0,10.0)
    private val ignoreWalls = setting("Ignore Walls", true)
    private val autoTool = setting("Auto Weapon", true)
    private val prefer = setting("Prefer", "Sword", listOf("Sword","Axe","None"))
    private val disableOnDeath = setting("Disable On Death", false)

    private var tickCounter = WaitCounter()
    private val msTimer = Timer()

    override fun onEnable() {
        //We don't want to lag
        tickCounter.addTick(20)
    }
    override fun onDisable() {
        msTimer.reset()
        tickCounter.reset()
    }

    override fun onTick() {
        if(shouldPause()) return
        if(delayMode.value == "Tick"){
            if(tickCounter.passed(tickDelay.value)) {
                loop()
                tickCounter.reset()
            } else
                tickCounter.addTick()
        }
    }

    @SubscribeEvent
    fun onRenderTick(event:TickEvent.RenderTickEvent){
        if(shouldPause()) return
        if(delayMode.value == "Ms"){
            if(msTimer.passed(msDelay.value.toDouble())) {
                loop()
                msTimer.reset()
            }
        }
    }

    private fun shouldPause():Boolean{
        if (mc.player == null || mc.player.isDead) {
            if (disableOnDeath.value) disable()
            return true
        }

        if (!whileEating.value) {
            val shield = mc.player.heldItemOffhand.getItem() == Items.SHIELD && mc.player.activeHand == EnumHand.OFF_HAND
            if (mc.player.isHandActive && !shield) {
                return true
            }
        }

        return false
    }

    private fun loop() {

        for (target in mc.world.loadedEntityList) {
            if (!EntityUtil.isLiving(target)) continue
            if (target === mc.player) continue
            if (mc.player.getDistance(target) > hitRange.value) continue
            if ((target as EntityLivingBase).health <= 0) continue
            if (target.hurtTime != 0) continue
            if (!ignoreWalls.value && !mc.player.canEntityBeSeen(target) && !canEntityFeetBeSeen(target)) continue

            if (attackPlayers.value && target is EntityPlayer && !FriendManager.isFriend(target)) {
                if (autoTool.value) equipBestWeapon(prefer.value)
                attack(target)
                if (!multiTarget.value) return
            } else {
                if (EntityUtil.mobTypeSettings(target, attackMobs.value, passive.value, neutral.value, hostile.value)) {
                    if (autoTool.value) equipBestWeapon(prefer.value)
                    attack(target)
                    if (!multiTarget.value) return
                }
            }
        }
    }

    private fun attack(e: Entity) {
        mc.playerController.attackEntity(mc.player, e)
        mc.player.swingArm(EnumHand.MAIN_HAND)
    }

    private fun canEntityFeetBeSeen(entityIn: Entity): Boolean {
        return mc.world.rayTraceBlocks(Vec3d(mc.player.posX, mc.player.posY + mc.player.getEyeHeight(), mc.player.posZ), Vec3d(entityIn.posX, entityIn.posY, entityIn.posZ), false, true, false) == null
    }

}